zcash_script/src/script/mod.rs

```diff
@@ -417,6 +439,30 @@ impl Evaluable for Code {
     }
 }
 
+impl Asm for Code {
+    fn to_asm(&self, attempt_sighash_decode: bool) -> String {
+        let mut v = Vec::new();
+        let is_unspendable = self.is_unspendable();
+        for r in self.parse_strict(&interpreter::Flags::StrictEnc) {
+            match r {
+                Ok(op) => v.push(op.to_asm(attempt_sighash_decode && !is_unspendable)),
+                Err(_) => {
+                    v.push("[error]".to_string());
+                    break;
+                }
+            }
+        }
+        v.join(" ")
```

Comment:
This didn’t come out quite as nicely as I had hoped, because it short-circuits while preserving the result thus far, but just an option.

```suggestion
        let is_unspendable = self.is_unspendable();
        self.parse(&interpreter::Flags::StrictEnc)
            .try_fold(
                String::new(),
                |mut s, r| r
                    .map(|op| s.push_str(op.to_asm(attempt_sighash_decode && !is_unspendable) + " "))
                    .map_err(|_| s.push_str("[error]")),
            )
            .unwrap_or_else(|x| x)
            .trim_end()
```

Reply:
That wasn't working and I couldn't make a similar idea work. I think it's confusing mixing functional and imperative code, but even if I changed it to concatenate strings in `try_fold()`, it doesn't solve the tricky part is that we need to stop it in the _next_  iteration after an error. The only idea I can come up is to return a flag along with the value in `try_fold()` to indicate that we should stop at the next iteration, but at that point the code would look much more confusing than the current imperative version.